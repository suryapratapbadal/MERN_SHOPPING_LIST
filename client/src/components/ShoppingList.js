import React, { Component } from 'react';
import {
    Container,
} from 'reactstrap';

import IconButton from '@material-ui/core/IconButton';

import { withStyles } from '@material-ui/core/styles';
import List from '@material-ui/core/List';
import ListItem from '@material-ui/core/ListItem';
import ListItemText from '@material-ui/core/ListItemText';
import EditIcon from '@material-ui/icons/Edit';
import SaveIcon from '@material-ui/icons/Save';
import Grid from '@material-ui/core/Grid';
import TextField from '@material-ui/core/TextField';
import ListItemSecondaryAction from '@material-ui/core/ListItemSecondaryAction';
import Checkbox from '@material-ui/core/Checkbox';
import LinearProgress from '@material-ui/core/LinearProgress';

import Recipe from './Recipe';
import AddRecipe from './AddRecipe';
import DeleteConfermation from './DeleteConfermation';
import ToggleMenu from './ToggleMenu';

import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';
import { getItems, updateItem } from '../actions/index';

export class ShoppingList extends Component {
    constructor(props) {
        super(props);
        this.state = {
            items: this.props.items,
            recipes: this.props.recipes,
            collapse: false,
            id: null,
            loader: true,
            editItem: false,
            editItemId: null,
            itemName: '',
            toggleDeleteConfirmation: false
        };
        this.editItem = this.editItem.bind(this);
        this.saveItem = this.saveItem.bind(this);
    }

    componentDidMount() {
        this.props.getItems();
    }

    componentDidUpdate(prevProps) {

        if (this.props.items !== prevProps.items)
            this.setState({ items: this.props.items, loader: false })
        if (this.props.recipes !== prevProps.recipes)
            this.setState({ recipes: this.props.recipes })
    }


    onUpdate = (id, data) => {
        this.props.updateItem(id, data)
    }


    toggle = id => {
        if (this.state.id === id && this.state.collapse === true) {
            this.setState({ collapse: false });
        } else if (this.state.id === id && this.state.collapse === false) {
            this.setState({ collapse: true });
        } else {
            this.setState({ collapse: true, id: id });
        }

    }

    toggleDeleteConfirmationDialogue = () => {
        this.setState({ toggleDeleteConfirmation: !this.state.toggleDeleteConfirmation })
    }
    onChange = event => {
        this.setState({ itemName: event.target.value });
    }

    editItem = (event, id) => {
        event.preventDefault();
        event.stopPropagation();
        this.setState({ editItem: true, editItemId: id, itemName: '' });
    }
    saveItem = (event, id) => {
        event.preventDefault();
        event.stopPropagation();
        if (this.state.itemName !== '') {
            this.onUpdate(id, { "name": this.state.itemName });
            this.setState({ editItem: false, editItemId: null });
        }
        else
            this.setState({ editItem: false, editItemId: null });
    }

    render() {
        const { items, loader } = this.state;
        const { classes, loading } = this.props;

        return (
            <Container className={classes.root}>
                {
                    loader ?
                        <LinearProgress /> :
                        <List
                            component="nav"
                        >
                            {loading && <LinearProgress />}
                            {items.map(({ _id, name, completed, recipes }) =>
                                (

                                    <Grid container key={_id} className={loading ? classes.listItemDeactivate: classes.listItem}>
                                        <Grid item xs={12}>
                                            <ListItem>
                                                <Grid item xs={1} >
                                                    <Checkbox
                                                        // onChange={this.onUpdate.bind(this, _id, { "completed": !completed })}
                                                        checked
                                                    />
                                                </Grid>
                                                <Grid item xs={6} className={completed ? classes.listItemDeactivate : ''}>
                                                    {
                                                        this.state.editItem && _id === this.state.editItemId ?
                                                            <TextField
                                                                id="standard-uncontrolled"
                                                                label="Item Name"
                                                                defaultValue={name}
                                                                className={classes.textField}
                                                                margin="normal"
                                                                onChange={this.onChange}
                                                                autoFocus
                                                            /> :
                                                            <ListItemText inset primary={name} />
                                                    }
                                                </Grid>
                                                <Grid item xs={1} className={completed ? classes.listItemDeactivate : ''}>
                                                    <IconButton
                                                        onClick={this.state.editItem && _id === this.state.editItemId ? event => this.saveItem(event, _id) : event => this.editItem(event, _id)}
                                                        style={{ marginRight: '1rem' }}
                                                        disableRipple>
                                                        {
                                                            this.state.editItem && _id === this.state.editItemId ?
                                                                <SaveIcon/> :
                                                                <EditIcon/>
                                                        }

                                                    </IconButton>
                                                </Grid>
                                                <Grid item xs={4} className={completed ? classes.listItemDeactivate : ''}>
                                                    <AddRecipe id={_id} style={{ float: 'right' }} />
                                                </Grid>

                                                <ListItemSecondaryAction>
                                                    <DeleteConfermation toggleDeleteConfirmation={this.state.toggleDeleteConfirmation} id={_id} />
                                                    <ToggleMenu toggleDeleteConfirmationDialogue={this.toggleDeleteConfirmationDialogue} completed={completed} id={_id} />
                                                </ListItemSecondaryAction>
                                            </ListItem>
                                        </Grid>
                                        <Grid item xs={12} className={completed ? classes.listItemDeactivate : ''}>
                                            <Recipe id={_id} recipes={recipes} />
                                        </Grid>
                                    </Grid>

                                )
                            )}
                        </List>
                }

            </Container >
        );
    }
}

const styles = theme => ({
    root: {
        width: '100%',

    },
    listItem: {
        marginBottom: '2rem',
        backgroundColor: theme.palette.background.paper,
    },
    listItemDeactivate: {
        opacity: 0.5,
        pointerEvents: 'none'
    },
    button: {
        margin: theme.spacing.unit,
    },
    textField: {
        marginLeft: theme.spacing.unit,
        marginRight: theme.spacing.unit,
        width: 150,
    },
    chip: {
        margin: theme.spacing.unit,
    },
});

export default connect(state => {
    const items = state.itemReducer.items || [];
    const recipes = state.itemReducer.recipes || [];
    const loading = state.itemReducer.loading || false;
    return {
        items,
        recipes,
        loading
    }
}, dispatch => {
    return bindActionCreators({ getItems: getItems, updateItem: updateItem }, dispatch)
}
)(withStyles(styles)(ShoppingList));