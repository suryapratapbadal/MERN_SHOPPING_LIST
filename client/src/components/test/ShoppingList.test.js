import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

// Component to be tested
import { ShoppingList } from '../ShoppingList';


const mockgetItemsfn = jest.fn();
const mockdeleteItemfn = jest.fn();
const mockupdateItemfn = jest.fn();
const items = [{ _id: '1', name: 'Milk' }];
const recipes = [];
const classes = { root: "ShoppingList-root-199", nested: "ShoppingList-nested-200", button: "ShoppingList-button-201", textField: "ShoppingList-textField-202" };

// Component Renders
describe('<ShoppingList />', () => {

  const wrapper = shallow(<ShoppingList items={items} recipes={recipes} loading={false} classes={classes} getItems={mockgetItemsfn} deleteItem={mockdeleteItemfn} updateItem={mockupdateItemfn}/>);
  wrapper.setState({ loader: false });
  const event = {
    stopPropagation: jest.fn(),
    target: {
      value: 'dummy'
    }
  }
  describe('render()', () => {

    test('renders the component', () => {

      expect(toJson(wrapper)).toMatchSnapshot();

    });

    test('componentDidMount()', () => {

      wrapper.instance().componentDidMount();
      expect(mockgetItemsfn).toHaveBeenCalled();

    });

    test('componentDidUpdate(prevProps)', () => {

      wrapper.instance().componentDidUpdate({});
      expect(wrapper.state().loader).toEqual(false);

    });


  });

  describe('onDeleteClick()', () => {

    test('works on delete button click', () => {
      wrapper.find('WithStyles(IconButton)').at(0).props().onClick(event);
      expect(mockdeleteItemfn).toBeCalledWith('1');
    });

  });

  describe('onClick()', () => {

    test('toggle the edit item state', () => {
      wrapper.find('pure(EditIcon)').at(0).props().onClick(event);
      expect(wrapper.state('editItem')).toEqual(true);
    });

  });

  describe('onChange()', () => {

    test('changes the checkbox state', () => {
      wrapper.find('WithStyles(Checkbox)').at(0).props().onChange('1',{'completed': true});
      expect(mockupdateItemfn).toBeCalledWith('1',{'completed': true});
    });

  });


  describe('onClick()', () => {

    test('toggle the collapse)', () => {
      wrapper.find('WithStyles(ListItem)').at(0).props().onClick('1');
      expect(wrapper.state('collapse')).toEqual(true);
      expect(wrapper.state('id')).toEqual('1');
    });

  });

});


