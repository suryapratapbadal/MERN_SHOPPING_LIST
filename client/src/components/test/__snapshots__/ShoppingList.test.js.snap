// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ShoppingList /> render() renders the component 1`] = `
<Container
  className="ShoppingList-root-199"
  tag="div"
>
  <WithStyles(List)
    component="nav"
  >
    <WithStyles(Grid)
      container={true}
      key="1"
    >
      <WithStyles(Grid)
        item={true}
        xs={12}
      >
        <WithStyles(ListItem)
          button={true}
          disabled={false}
          onClick={[Function]}
        >
          <WithStyles(Grid)
            item={true}
            xs={1}
          >
            <WithStyles(ListItemIcon)>
              <WithStyles(IconButton)
                aria-label="Delete"
                color="secondary"
                disableRipple={true}
                onClick={[Function]}
                style={
                  Object {
                    "marginRight": "1rem",
                  }
                }
              >
                <pure(DeleteIcon) />
              </WithStyles(IconButton)>
            </WithStyles(ListItemIcon)>
          </WithStyles(Grid)>
          <WithStyles(Grid)
            item={true}
            xs={5}
          >
            <WithStyles(ListItemText)
              inset={true}
              primary="Milk"
            />
          </WithStyles(Grid)>
          <WithStyles(Grid)
            item={true}
            xs={1}
          >
            <WithStyles(IconButton)
              disableRipple={true}
              style={
                Object {
                  "marginRight": "1rem",
                }
              }
            >
              <pure(EditIcon)
                onClick={[Function]}
              />
            </WithStyles(IconButton)>
          </WithStyles(Grid)>
          <WithStyles(Grid)
            item={true}
            xs={4}
          >
            <Connect(WithStyles(AddRecipe))
              item_id="1"
              style={
                Object {
                  "float": "right",
                }
              }
            />
          </WithStyles(Grid)>
          <WithStyles(Grid)
            item={true}
            xs={1}
          >
            <pure(ExpandMoreIcon) />
          </WithStyles(Grid)>
          <WithStyles(ListItemSecondaryAction)>
            <WithStyles(Checkbox)
              onChange={[Function]}
            />
          </WithStyles(ListItemSecondaryAction)>
        </WithStyles(ListItem)>
      </WithStyles(Grid)>
      <WithStyles(Grid)
        item={true}
        xs={12}
      >
        <Connect(WithStyles(Recipe))
          collapse={false}
          id="1"
          open_id={null}
        />
      </WithStyles(Grid)>
    </WithStyles(Grid)>
  </WithStyles(List)>
</Container>
`;
